pkg/logbridge/logbridge.go

```diff
@@ -0,0 +1,88 @@
+package logbridge
+
+import (
+	"bufio"
+	"io"
+
+	"github.com/square/p2/pkg/logging"
+)
+
+// This type is agnostic of pipes. os.File conforms to io.Reader and io.Writer
+// so it should not be difficult to migrate
+type Bridge struct {
+	src  io.Reader
+	dest io.Writer
+
+	lines         chan []byte
+	needsTeardown bool
+}
+
+// New takes two files src and dest and returns a composite object of the
+// two.
+func New(src io.Reader, dest io.Writer, capacity int) Bridge {
+	if src == nil {
+		panic("Argument Error: Reader cannot be nil")
+	}
+	if dest == nil {
+		panic("Argument Error: Reader cannot be nil")
+	}
+
+	return Bridge{
+		src:           src,
+		dest:          dest,
+		lines:         make(chan []byte, capacity),
+		needsTeardown: true,
+	}
+}
+
+// Run is responsible for reading from src and writing to dst until src is
+// consumed or an error is encountered. The caller is responsible for closing
+// and flushing the reader and writer.
+func (b *Bridge) Run() error {
+	var buf []byte
+	var moreLines bool
+	go b.blockingRead()
+
+	for {
+		select {
+		case buf, moreLines = <-b.lines:
+			if !moreLines {
+				return nil
+			}
+			_, err := b.dest.Write(buf)
+			if err != nil {
+				logging.DefaultLogger.WithError(err).Error("logbridge: encountered fatal error during write")
+				b.teardown()
+				return err
+			}
+		default:
+		}
+	}
+
+	return nil
+}
+
+func (b *Bridge) blockingRead() {
+	scanner := bufio.NewScanner(b.src)
+	for scanner.Scan() {
+		if err := scanner.Err(); err != nil {
+			logging.DefaultLogger.WithError(err).Errorln("Encountered error while reading from src. Proceeding.")
+			break
+		}
+
+		select {
+		case b.lines <- scanner.Bytes():
+		default:
+			logging.DefaultLogger.NoFields().Debugln("Overflow encountered. Dropping this message.")
+			continue
+		}
+	}
+
+	b.teardown()
+}
+
+func (b *Bridge) teardown() {
+	if b.needsTeardown {
```

Comment:
b.needsTeardown never gets set to false. Although frankly I would ditch this function altogether. I recommend not sharing the send side of a channel across scopes or goroutines anyway, unless very carefully coordinated to avoid panic.